### PR TITLE
Ticket 12264 cache file engine compression

### PR DESCRIPTION
**Pull request by Erwane**
Cache FileEngine compress files before write.
from 50% to 90% file size saved

### CODE REVIEW

File: tests/TestCase/Cache/Engine/FileEngineTest.php [hunk 1]
```diff
@@ -641,4 +641,32 @@ public function testAdd()
...
+
+    /**
+     * Test reading and writing to the compressed cache.
+     *
+     * @return void
+     * @expectedException \PHPUnit\Framework\Error\Error
+     */
+    public function testReadAndwriteCompressed()
+    {
+        $this->_configCache(['compression' => true]);
```
[the remark this answers is not in the document]
Comment (josbeir) [reply tests/TestCase/Cache/Engine/FileEngineTest.php]: Thats what i was wondering as well as the goal of caching is to reduce system load for one, how does this impact performance. Personally i feel like this is a feature that should be handled by a custom plugin or something.

[the remark this answers is not in the document]
Comment (Erwane) [reply tests/TestCase/Cache/Engine/FileEngineTest.php]: i'l try to do benchmark.
The things is : CPU handle very well the compression now, specially on serialized data (text).
It's a very low cost i think and sometime, a huge gain of Bytes on SSD.
In my project, some caches are compressed, some not, depends of the content but from 400k to 40k * 10'000 files ...

[the remark this answers is not in the document]
Comment (markstory) [reply tests/TestCase/Cache/Engine/FileEngineTest.php]: If high performance is a requirement why are you using files for caching?

[the remark this answers is not in the document]
Comment (Erwane) [reply tests/TestCase/Cache/Engine/FileEngineTest.php]: i have tons of cache files and it's enought fast for me (SSD).
My problem is the size on the ssd.

**Comment by dereuromark** [comment]
ping @Erwane

File: tests/TestCase/Cache/Engine/FileEngineTest.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (dereuromark) [reply tests/TestCase/Cache/Engine/FileEngineTest.php]: I agree that compression would mainly make sense if the cache would be on a different system, thus bandwidth would be greatly improving the speed of writing and reading then.
This indeed could be a custom extension as plugin solution. Maybe we can close this for core level.

**Comment by Erwane** [comment]
I've made this, a week ago but on a closed issue : 

https://gist.github.com/Erwane/dd3bf45671c23cd1d2091a8f34f65541

It's a "Processor" options for all cache. Need more tests but it's aonly a Proof of Concept ;)

**Pull request closed without merging**